app/channel.go

```diff
@@ -44,71 +44,54 @@ func (a *App) JoinDefaultChannels(teamId string, user *model.User, shouldBeAdmin
 		}
 	}
 
-	if result := <-a.Srv.Store.Channel().GetByName(teamId, "town-square", true); result.Err != nil {
-		err = result.Err
-	} else {
-		townSquare := result.Data.(*model.Channel)
-
-		cm := &model.ChannelMember{
-			ChannelId:   townSquare.Id,
-			UserId:      user.Id,
-			SchemeUser:  true,
-			SchemeAdmin: shouldBeAdmin,
-			NotifyProps: model.GetDefaultChannelNotifyProps(),
-		}
-
-		if cmResult := <-a.Srv.Store.Channel().SaveMember(cm); cmResult.Err != nil {
-			err = cmResult.Err
-		}
-		if result := <-a.Srv.Store.ChannelMemberHistory().LogJoinEvent(user.Id, townSquare.Id, model.GetMillis()); result.Err != nil {
-			mlog.Warn(fmt.Sprintf("Failed to update ChannelMemberHistory table %v", result.Err))
-		}
-
-		if *a.Config().ServiceSettings.ExperimentalEnableDefaultChannelLeaveJoinMessages {
-			if requestor == nil {
-				if err := a.postJoinTeamMessage(user, townSquare); err != nil {
-					mlog.Error(fmt.Sprint("Failed to post join/leave message", err))
-				}
-			} else {
-				if err := a.postAddToTeamMessage(requestor, user, townSquare, ""); err != nil {
-					mlog.Error(fmt.Sprint("Failed to post join/leave message", err))
-				}
-			}
-		}
+	defaultChannelList := []string{"town-square"}
 
-		a.InvalidateCacheForChannelMembers(result.Data.(*model.Channel).Id)
+	if len(a.Config().TeamSettings.ExperimentalDefaultChannels) == 0 {
+		defaultChannelList = append(defaultChannelList, "off-topic")
+	} else {
+		defaultChannelList = append(defaultChannelList, a.Config().TeamSettings.ExperimentalDefaultChannels...)
 	}
 
-	if result := <-a.Srv.Store.Channel().GetByName(teamId, "off-topic", true); result.Err != nil {
-		err = result.Err
-	} else if offTopic := result.Data.(*model.Channel); offTopic.Type == model.CHANNEL_OPEN {
+	for _, channelName := range defaultChannelList {
+		if result := <-a.Srv.Store.Channel().GetByName(teamId, channelName, true); result.Err != nil {
+			err = result.Err
+		} else {
 
-		cm := &model.ChannelMember{
-			ChannelId:   offTopic.Id,
-			UserId:      user.Id,
-			SchemeUser:  true,
-			SchemeAdmin: shouldBeAdmin,
-			NotifyProps: model.GetDefaultChannelNotifyProps(),
-		}
+			channel := result.Data.(*model.Channel)
 
-		if cmResult := <-a.Srv.Store.Channel().SaveMember(cm); cmResult.Err != nil {
-			err = cmResult.Err
-		}
-		if result := <-a.Srv.Store.ChannelMemberHistory().LogJoinEvent(user.Id, offTopic.Id, model.GetMillis()); result.Err != nil {
-			mlog.Warn(fmt.Sprintf("Failed to update ChannelMemberHistory table %v", result.Err))
-		}
+			if channelName == "off-topic" && channel.Type != model.CHANNEL_OPEN {
```

Comment:
@jwilander Should we implement that behavior for all default channels? I was a bit uncertain here, on one hand the admin explicitly set the channel to a default channel, on another hand we should maybe only allow open channels to be default. The current implementation mimics the pre-PR code base. Happy to change it.

Reply:
Hmmm good question. I'm leaning toward only public channels can be default channels. Thoughts @jasonblais ?